### PR TITLE
Future deserialization without available client

### PR DESCRIPTION
**Pull request by fjetter**
This is similar to 

This grew a bit more complicated because I stumbled over https://github.com/dask/distributed/issues/7498 again and had a deeper look and tried to preserve the "pass futures in collections" feature but it is fundamentally flawed. While I could bypass most of the "accidental client creations" in this PR (which is good), the fundamental flaw about it being possible to release a future before it is being deserialized is still there and hard to avoid without a more fundamental approach.

Two notable changes

- The get_worker function now actually does what it is claiming. It returns the worker of the currently running _task_. Most notably this will raise if being executed outside of the threadpool. This is to a certain degree inconvenient since deserialization does typically not happen inside of our thread but I consider this important considering that our entire test suite is running asynchronously.
- Futures, etc. are actually only initializing a client lazily if a client (or worker) isn't immediately available. This avoids problems about "when, where and how" do we deserialize a task, e.g. we can deserialize a Future or Queue object on the event loop but since we're only interacting with it in the threadpool, the initialization will be deferred. To a certain degree this is also nicer for diagnostics but that's not the motivating change for this.

This is again a supporting PR for https://github.com/dask/distributed/pull/7564

### CODE REVIEW

File: distributed/worker.py [hunk 1]
```diff
@@ -3190,16 +3198,18 @@ def apply_function_actor(
     with active_threads_lock:
         active_threads[ident] = key
 
-    thread_state.execution_state = execution_state
-    thread_state.key = key
-    thread_state.actor = True
-
-    result = function(*args, **kwargs)
```
Comment (fjetter) [inline distributed/worker.py:3197]: State was actually leaking all over the place making get_worker in our test suite extremely unreliable.

File: distributed/worker.py [hunk 2]
```diff
@@ -2719,10 +2724,7 @@ def get_worker() -> Worker:
     try:
         return thread_state.execution_state["worker"]
     except AttributeError:
-        try:
-            return first(w for w in Worker._instances if w.status in WORKER_ANY_RUNNING)
-        except StopIteration:
-            raise ValueError("No workers found")
```
Comment (fjetter) [inline distributed/worker.py:2725]: I get the appeal of this but this can throw off some of our test logic and tests suggest that certain functionality would work which in reality doesn't.

File: distributed/tests/test_multi_locks.py [hunk 3]
```diff
@@ -56,8 +56,8 @@ async def test_timeout(c, s, a, b):
     await lock1.release()
 
 
-@gen_cluster()
-async def test_timeout_wake_waiter(s, a, b):
+@gen_cluster(client=True)
+async def test_timeout_wake_waiter(c, s, a, b):
```
Comment (fjetter) [inline distributed/tests/test_multi_locks.py:60]: The change to get_worker causes for instance these things to break. On main, this test is actually not allowed to pass since there is no worker and no client in the context for the multi lock. However, due to the "pick first worker in the process" logic, it would pick a random worker and create implicitly a worker client.
In most cases this is not a big deal but implicit, surprising client creations can cause things like https://github.com/dask/distributed/issues/7498

**Comment by fjetter** [comment]
Interesting. I seem to hit a **related** recursion error in `test_profile_server`... 🤯

File: distributed/worker.py (hunk 2 above)
Comment (fjetter) [reply distributed/worker.py:2725]: i.e. due to our async tests, we are likely to get a worker just because in the same process an instance already exists. However, in reality there is no valid worker in the context, i.e. if the test example would be executed on a "real" cluster, it would fail

**Comment by fjetter** [comment]
OK, so I'm not entirely sure if the profiling thing is indeed related but apparently with this change, we're more likely to generate a very deeply nested profile message. The profile collection works out nicely but as soon as one tries to submit this profile message, we're hitting a recursion error during serialization.

We can't truly protect ourselves from this during serialization since the serialization code does not know what is OK to cut out and what isn't. The only way I see we can deal with this is to be more conservative with the cutoff in profiling. Local testing found ~250 stacks to be the magical limit to get `test_profile_server` pass.

File: distributed/profile.py [hunk 4]
```diff
@@ -155,15 +155,20 @@ def process(
     merge
     """
     if depth is None:
-        depth = sys.getrecursionlimit() - 50
-    if depth <= 0:
-        return None
```
Comment (fjetter) [inline distributed/profile.py:160]: this check caused us to not collect any information in these cases. I think it's still valuable to get a snapshot through even if it's not the lowest frame. Moving the depth check further below achieves this (see test_profile.py)

File: distributed/protocol/tests/test_protocol.py [hunk 5]
```diff
@@ -393,3 +394,21 @@ def _(header, frames):
...
+            depth -= 1
+            d["children"] = d = {}
+        return msg
+
+    msg = gen_deeply_nested(sys.getrecursionlimit() - 100)
+    with pytest.raises(TypeError, match="Could not serialize object"):
+        serialize(msg, on_error="raise")
+
+    msg = gen_deeply_nested(sys.getrecursionlimit() // 4)
+    assert isinstance(serialize(msg), tuple)
```
Comment (fjetter) [inline distributed/protocol/tests/test_protocol.py:414]: This test logic is not super useful but it shows that we are hitting such an error, we're raising/ignoring it according to `on_error` and if the stack size is large but not too large, everything passes. So all in all this is an OK-ish test but not great

File: distributed/shuffle/tests/test_shuffle.py [hunk 6]
```diff
@@ -74,10 +74,10 @@ async def clean_scheduler(
...
 
 
-@pytest.mark.skipif(
-    pa is not None,
-    reason="We don't have a CI job that is installing a very old pyarrow version",
-)
+# @pytest.mark.skipif(
+#     pa is not None,
+#     reason="We don't have a CI job that is installing a very old pyarrow version",
+# )
```
Comment (hendrikmakait) [inline distributed/shuffle/tests/test_shuffle.py:80]:
```suggestion
```

File: distributed/shuffle/tests/test_shuffle.py [hunk 7]
```diff
@@ -86,8 +86,9 @@ async def test_minimal_version(c, s, a, b):
         dtypes={"x": float, "y": float},
         freq="10 s",
     )
-    with pytest.raises(RuntimeError, match="requires pyarrow"):
-        await c.compute(dd.shuffle.shuffle(df, "x", shuffle="p2p"))
+    # with pytest.raises(RuntimeError, match="requires pyarrow"):
```
Comment (hendrikmakait) [inline distributed/shuffle/tests/test_shuffle.py:89]:
```suggestion
```

**Review by hendrikmakait — APPROVED** [review]
The code generally looks good to me. I have a few minor nits, and a few edge cases are untested. I suggest marking them with `pragma: nocover` if we don't want to test these.

File: distributed/shuffle/tests/test_shuffle.py (hunk 6 above)
Comment (jrbourbeau) [reply distributed/shuffle/tests/test_shuffle.py]: It's clear to me why this change was made to this test. My impression is that what's currently on `main` is still want we want. I've reverted back to what's on `main` (confirmed the test passes locally when `pyarrow` isn't installed). Happy to submit a follow-up PR if needed

**Comment by jrbourbeau** [comment]
Noting test failures are unrelated to the changes in this PR (xref https://github.com/dask/distributed/issues/7688)

**Comment by martindurant** [comment]
Since this PR, the following code fails surprisingly:
```
>>> client = distributed.Client()
>>> client.run(distributed.worker.get_worker)
ValueError: No worker found
```

This got noticed because it was causing fsspec's dask implementation to hang in tests.

**Comment by martindurant** [comment]
I did [this](https://github.com/fsspec/filesystem_spec/pull/1224/commits/d003d7f0a2ed1a2361fdd8a8d3791287b60a9d15#diff-206aac57bbb0452d25979c0f144be048b8a6767274b3cd5a10af4f8245c59ceaR20-R22) to workaround, since my filesystem only needs to know whether this is a process housing a worker, not if it happens to be running in a task.

**Comment by jrbourbeau** [comment]
Thanks @martindurant, this looks related to https://github.com/dask/distributed/issues/7696. There's a good conversation in that issue, but https://github.com/dask/distributed/issues/7696#issuecomment-1479657941 is the specific comment that relates to the snippet you posted

**Comment by martindurant** [comment]
Thanks for the pointer. From fsspec's point of view, the question is "am I on a worker machine", independent of serialisation or task state, so I'll use the simple approach.
